File: backend/firebase.js
require("dotenv").config();

const { initializeApp, cert } = require("firebase-admin/app");
const { getFirestore } = require("firebase-admin/firestore");
const { getAuth } = require("firebase-admin/auth");

let credentialConfig;

if (process.env.FIREBASE_SERVICE_ACCOUNT) {
  try {
    const serviceAccountConfig = JSON.parse(
      process.env.FIREBASE_SERVICE_ACCOUNT,
    );
    credentialConfig = cert(serviceAccountConfig);
    console.log(
      "🔥 Firebase initialized via secure cloud environment variable.",
    );
  } catch (error) {
    console.error(
      "CRITICAL: Failed to parse FIREBASE_SERVICE_ACCOUNT environment variable.",
      error,
    );
    process.exit(1);
  }
} else {
  try {
    const serviceAccount = require("./firebase-service-account.json");
    credentialConfig = cert(serviceAccount);
    console.log("🔥 Firebase initialized via local service account file.");
  } catch (error) {
    console.error(
      "CRITICAL: Local firebase-service-account.json not found.",
      error,
    );
    process.exit(1);
  }
}

initializeApp({
  credential: credentialConfig,
});

const db = getFirestore();
const auth = getAuth();

module.exports = { db, auth };

File: backend/firebase.py
import json
import os
import sys

import firebase_admin
from dotenv import load_dotenv
from firebase_admin import auth as firebase_auth
from firebase_admin import credentials, firestore

load_dotenv()

service_account_env = os.environ.get("FIREBASE_SERVICE_ACCOUNT")

if service_account_env:
    try:
        cred = credentials.Certificate(json.loads(service_account_env))
        print("🔥 Firebase initialized via secure cloud environment variable.")
    except Exception as error:
        print(
            "CRITICAL: Failed to parse FIREBASE_SERVICE_ACCOUNT environment variable.",
            error,
        )
        sys.exit(1)
else:
    try:
        cred = credentials.Certificate("firebase-service-account.json")
        print("🔥 Firebase initialized via local service account file.")
    except Exception as error:
        print("CRITICAL: Local firebase-service-account.json not found.", error)
        sys.exit(1)

firebase_admin.initialize_app(cred)

db = firestore.client()
auth = firebase_auth
